@here/harp-mapview/lib/geometry/TileGeometryLoader.ts

```diff
@@ -0,0 +1,170 @@
+    private m_isFinished: boolean = false;
+
+    constructor(private m_tile: Tile) {}
+
+    get tile(): Tile {
+        return this.m_tile;
+    }
+
+    get isFinished(): boolean {
+        return this.m_isFinished;
+    }
+
+    get basicGeometryLoaded(): boolean {
+        return this.m_tile.hasGeometry;
+    }
+
+    get allGeometryLoaded(): boolean {
+        return this.m_isFinished;
+    }
+
+    setDecodedTile(decodedTile: DecodedTile) {
+        this.m_decodedTile = this.m_tile.decodedTile;
+    }
+
+    update(): void {
+        if (this.m_decodedTile === undefined && this.m_tile.decodedTile !== undefined) {
+            this.setDecodedTile(this.m_tile.decodedTile);
+            this.prepareForRender();
+            this.finish();
+        }
+    }
+
+    dispose(): void {
+        this.m_decodedTile = undefined;
+    }
+
+    private finish() {
+        this.m_tile.removeDecodedTile();
+        this.m_isFinished = true;
+    }
+
+    /**
+     * Called by [[VisibleTileSet]] to mark that [[Tile]] is visible and it should prepare geometry.
+     */
+    private prepareForRender() {
+        // If the tile is not ready for display, or if it has become invisible while being loaded,
+        // for example by moving the camera, the tile is not finished and its geometry is not
+        // created. This is an optimization for fast camera movements and zooms.
+        const tile = this.tile;
+        const decodedTile = this.m_decodedTile;
+        this.m_decodedTile = undefined;
+        if (decodedTile === undefined || tile.disposed || !tile.isVisible) {
+            return;
+        }
+        setTimeout(() => {
+            const stats = PerformanceStatistics.instance;
+            // If the tile has become invisible while being loaded, for example by moving the
+            // camera, the tile is not finished and its geometry is not created. This is an
+            // optimization for fast camera movements and zooms.
+            if (!tile.isVisible) {
+                // Dispose the tile from the visible set, so it can be reloaded properly next time
+                // it is needed.
+                tile.mapView.visibleTileSet.disposeTile(tile);
+
+                if (stats.enabled) {
+                    stats.currentFrame.addMessage(
+                        `Decoded tile: ${tile.dataSource.name} # lvl=${tile.tileKey.level} col=${
+                            tile.tileKey.column
+                        } row=${tile.tileKey.row} DISCARDED - invisible`
+                    );
+                }
+                return;
+            }
+            let now = 0;
+            if (stats.enabled) {
+                now = PerformanceTimer.now();
+            }
+
+            const geometryCreator = new TileGeometryCreator();
+
+            // Speedup and simplify following code: Test all techniques if they intersect with the
+            // disabledKinds, in which case they are flagged. The disabledKinds can be ignored
```

Comment:
disabledKinds is some concept to be introduced in another loader?

Reply:
Removed the comment. That one comes later.